### PR TITLE
SearchIO HHsuite parser

### PR DESCRIPTION
**Pull request by linucks**
This pull request adds a SearchIO parser for the HH-suite output format as described here:

https://github.com/soedinglab/hh-suite/wiki

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

*Edit: Fixed the check boxes, Peter*

### CODE REVIEW

**Comment by linucks** [comment]
The flake8 test is currently failing with:

Bio/SearchIO/HHsuiteIO/hhsuite2_text.py:139:1: RST301 Unexpected indentation.
Bio/SearchIO/HHsuiteIO/hhsuite2_text.py:140:1: RST201 Block quote ends without a blank line; unexpected unindent.

This is because in a block comment I have a block of text showing what is being parsed so that someone can better understand what the code is doing. Is there a way I can keep a similar comment and not have the test fail?

**Comment by peterjc** [comment]
You probably need to use a literal block (end previous line with double colon):

http://docutils.sourceforge.net/docs/user/rst/quickref.html#literal-blocks

Have an experiment in an RST editor, e.g. http://rst.ninjs.org/

**Comment by linucks** [comment]
That's great - thanks for the tip @peterjc

File: Bio/SearchIO/HHsuiteIO/hhsuite2_text.py [hunk 1]
```diff
@@ -210,9 +210,9 @@ def _create_qresult(self, hit_blocks):
         return [qresult]
 
     def _unique_hit_id(self, hit_id, existing_ids, separator='_'):
-        """Return a unique hit id.(PRIVATE).
+        """Return a unique hit id. (PRIVATE).
```
Comment (peterjc) [inline Bio/SearchIO/HHsuiteIO/hhsuite2_text.py]: Minor, but you can remove the full stop after "id" here.

File: Bio/SearchIO/HHsuiteIO/hhsuite2_text.py [hunk 2]
```diff
@@ -132,15 +132,16 @@ def _parse_hit_block(self):
...
-        Parses block ala:
-        Q ss_pred             ceecchHHHHHHHHHHHHHHHHHHHhhhhhcCCCCccc
-        Q 4P79:A|PDBID|C  160 YELGPALYLGWSASLLSILGGICVFSTAAASSKEEPAT  197 (198)
-        Q Consensus       160 ~~~g~sf~l~~~~~~l~~~~~~l~~~~~~~~~~~~~~~  197 (198)
-                              .++|||||++|++.++.+++++++++..+..++++..+
-        T Consensus       327 ~~~GwS~~l~~~s~~l~lia~~l~~~~~~~~~~~~~~~  364 (364)
-        T 5B2G_A          327 REMGASLYVGWAASGLLLLGGGLLCCSGPSSGENLYFQ  364 (364)
-        T ss_dssp             EEECTHHHHHHHHHHHHHHHHHHHHCC-----------
-        T ss_pred             cccchHHHHHHHHHHHHHHHHHHHHhcCCCCCCccccC
+        Parses block such as ::
```
Comment (peterjc) [inline Bio/SearchIO/HHsuiteIO/hhsuite2_text.py]: You can remove the space before the colons. I would also rephrase, maybe:

```
Parses an alignment block like this::
```

**Comment by linucks** [comment]
I've just realised I may have misunderstood the object model.

Taking the example file Tests/HHsuite/2uvo_hhblits.hhr there are 2 hits from the pdb 2wga, one at position 2, the other at position 6. I'm currently creating 2 hits, one with the id 2wga_1 and the other with the id 2wga_2. I'm now guessing that this is one hit with 2 hsp's? If so how would a user get the entire list of hsps in the order they were parsed from the file as I can't see a way to do this?

Also, I'd like to raise some exceptions in case I encounter unexpected output in the file. What's the best exception to raise? Looking at other files it seems that it's either ValueError or RuntimeError - is one preferred?

**Comment by bow** [comment]
Hi @linucks,

Thanks for the PR! I have not finished reviewing, but I can already answer some of your questions:

> Taking the example file Tests/HHsuite/2uvo_hhblits.hhr there are 2 hits from the pdb 2wga, one at position 2, the other at position 6. I'm currently creating 2 hits, one with the id 2wga_1 and the other with the id 2wga_2. I'm now guessing that this is one hit with 2 hsp's? If so how would a user get the entire list of hsps in the order they were parsed from the file as I can't see a way to do this?

Indeed it is more correct to store this as 2 HSPs in the current object model. The model itself does not impose any sorting order. You may be able to use something like [`groupby`](https://docs.python.org/3.7/library/itertools.html#itertools.groupby), if you are parsing the raw table lines first. But, seeing that  the input file the HSPs seem to be sorted by any of the `Prob` / `E-value` / `P-value` / `Score` values in descending order, perhaps creating them HSP objects first and then sorting by any of these values is possible?

> Also, I'd like to raise some exceptions in case I encounter unexpected output in the file. What's the best exception to raise? Looking at other files it seems that it's either ValueError or RuntimeError - is one preferred?

`ValueError` is preferred. `RuntimeError` is only used in BlatIO in SearchIO. In retrospect, that is even probably better as `ValueError`.

Hope this helps!

**Comment by linucks** [comment]
Hi @bow

Thanks for the quick reply. I'll push some changes so that I work better within the model. However, I'm not sure I completely understand what you're saying. Most of the use cases I have for working with a SearchIO model involve getting, for example, the top 3 HSPs (across all hits) by bitscore or e-value. Once the HSPs are  within the SearchIO model, is there no easy way for a user to get the HSPs from the QueryResult sorted by a particular attribute?

**Comment by bow** [comment]
Ah, indeed there is no one, single method call from the `QueryResult` instance for that use case. But you can return all the `HSP`s from a single `QueryResult` using the [`QueryResult.hsps` property](
https://github.com/biopython/biopython/blob/master/Bio/SearchIO/_model/query.py#L456), which you can then sort by an HSP attribute. Something like:

```
>>> hsps = sorted(query.hsps, key=lambda hsp: hsp.prob)
```

The `sorted` call is necessary because otherwise you would still get the `HSP`s in the order of the `Hit` objects stored in `QueryResult`.

So not the shortest function call, but would this be enough for your use case?

**Comment by linucks** [comment]
Hi @bow,

That would work in most cases and is probably suitable for what I need, but I was wondering if there should be a dedicated function in the API that supports this?

To me the parser isn't doing it's job properly if it doesn't return the HSPs in the order they were output by the program (HHsearch/HHBlits in this case), as that's why the user ran the program; to get a list of HSPs in the order the program thinks are most suitable. Returning the HSPs ranked by a user-selected score may not do this though. For example in my test case Tests/HHsuite/allx.hhr as the query was effectively nonsense ('XXXXX'), the top 4 hits returned by HHBlits all have the same Probability, Score and E-value so sorting by (e.g.) E-value wouldn't return the original order. Although this case is somewhat contrived, I could well imagine where a program had 2 top hits with the same probability, but different E-values, but it shouldn't be up to the user to understand how the program sorts it's hits internally and know what to do to get the original order.

Would it therefore make sense to assign all HSPs a global index or rank within the QueryResult based on the order they were output by the program, and have it so that the default sorting order when calling query.hsps returned the HSPs in that order?

**Comment by bow** [comment]
@linucks, thanks for the thorough explanation. I have a better idea about your use case now.

I think it is fine if we add a standard attribute to each HSP object to track the original output order. This attribute has not been necessary for the existing parsers, as they sort their outputs by hit first, then by HSP (so the same as the default sort order we use). Of course, this does not really capture output formats that sort by HSP only.

Maybe the new HSP attribute should be called `parse_rank`?. The default value should be `-1`, to guarantee that `sorted` always works even when parsers do not store this value (as `sorted()` works with `None` in Python 2.7, but not in Python 3.x). Also, since [`sorted`](https://docs.python.org/3/library/functions.html#sorted) is also stable, this won't introduce any breaking changes with the `QueryResult.hsps` return values of the existing parsers.

@peterjc, would you prefer to have this change in this PR or shall we make a new one first?

**Comment by linucks** [comment]
Ok @bow. I've now refactored to store everything within the hits, which is much neater!

**Comment by linucks** [comment]
It looks like the AppVeyor build failed with "Fallback clouds stack is too big." I'm guessing this is an infrastructure issue and not something to do with the code? Is there anything I need to do about this or am I ok just ignore it?

@bow I was thinking about the HSP attribute and am not sure about 'parse_rank', as to me that seems too closely related to how the data was processed, which should be irrelevant to a user (the code may have parsed the file backwards for example). Just 'rank' seems a good choice to me as it describes the canonical ordering of the HSPs and is not related to anything else. What do you think?

**Comment by peterjc** [comment]
Applying #1987 and then rebasing this work sounds sensible.

It looks like AppVeyor is continuing to have difficulties - right now https://status.appveyor.com/ looks fine, but the *"Fallback clouds stack is too big."* message was reported earlier this month:

https://help.appveyor.com/discussions/problems/22315-builds-are-failing-most-of-the-time-fallback-clouds-stack-is-too-big

File: Tests/test_SearchIO_hhsuite2_text.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        hit = qresult[0]
+        self.assertEqual('2uvo_A', hit.id)
+        self.assertEqual('Agglutinin isolectin 1; carbohydrate-binding protein, hevein domain, chitin-binding,'
+                         ' GERM agglutinin, chitin-binding protein; HET: NDG NAG GOL; 1.40A {Triticum aestivum}'
+                         ' PDB: 1wgc_A* 2cwg_A* 2x3t_A* 4aml_A* 7wga_A 9wga_A 2wgc_A 1wgt_A 1k7t_A* 1k7v_A* 1k7u_A'
+                         ' 2x52_A* 1t0w_A*', hit.description)
+        self.assertTrue(hit.is_included)
+        self.assertEqual(3.7e-34, hit.evalue)
+        self.assertEqual(210.31, hit.score)
+#         self.assertEqual(0.5, hit.bias)
```
[the remark this answers is not in the document]
Comment (linucks) [reply Tests/test_SearchIO_hhsuite2_text.py]: Addressed with commit ded6d78a2a7751fbffa13461739c9b9b4b20e572

File: Tests/test_SearchIO_hhsuite2_text.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        self.assertEqual(3.4E+04, hit.evalue)
+        self.assertEqual(-0.01, hit.score)
+        self.assertEqual(1, len(hit))
+
+        hsp = hit.hsps[0]
+        self.assertTrue(hsp.is_included)
+        self.assertEqual(3.4E+04, hsp.evalue)
+        self.assertEqual(-0.01, hsp.score)
+        self.assertEqual(0.04, hsp.prob)
+        self.assertEqual(24, hsp.hit_start)
```
[the remark this answers is not in the document]
Comment (linucks) [reply Tests/test_SearchIO_hhsuite2_text.py]: Addressed with commit ded6d78a2a7751fbffa13461739c9b9b4b20e572

**Comment by linucks** [comment]
Hi @bow I've pushed some commits that should address everything you raised.

I've just noticed that HHSUITE have released a new version (3) of their software. The output is identical  to version 2, so I guess it should be made clear that the parser supports both versions 2 and 3. It won't support version 1, however, as the format is completely different - and this is probably not relevant as it's long deprecated.

What's the best way of doing this?

**Comment by bow** [comment]
Thanks for the updates @linucks, it looks like after integrating the changes in #1987 and storing the original sort order, this can be merged :+1: .

As for indicating compatibility, in addition to mentioning this in `Bio.SearchIO.HHsuiteIO.__init__.py`, I think adding a new `hhsuite3-text` key [here](https://github.com/linucks/biopython/blob/searchio_hhsuite_parser/Bio/SearchIO/__init__.py#L230) would be ok. The value of the key can be set initially to be the same as `hhsuite2-text`'s and it can be updated in case the output for version 3 starts to deviate from version 2.

**Comment by linucks** [comment]
That's great - thanks @bow ! Sorry to have to ask, but how do I integrate the changes in #1987? Do I need to pull in a specific set of changes? If you could let me know the relevant git commands, that would be greatly appreciated.

**Comment by bow** [comment]
@linucks, we could do a `git rebase` with this:

```
$ git checkout master 
$ git pull upstream master   # update to the latest master
$ git checkout searchio_hhsuite_parser
$ git rebase -i master  # this opens up an interactive window, and should show your commits
$ git push origin searchio_hhsuite_parser --force
```

A rebase is not what everyone does as it changes the commit hashes (since we are using a different commit 'base' for your changeset), hence the need for `--force`. But it does keep the history linear and easier to search through, which is why we do it.

If you prefer, I also don't mind if you create a new branch from your current one and then open a new PR on that branch:
```
$ git checkout master 
$ git pull upstream master   # update to the latest master
$ git checkout searchio_hhsuite_parser
$ git checkout -b new_branch
$ git rebase -i master  # this opens up an interactive window, and should show your commits
$ git push origin new_branch
```

In any case, you would of course need to commit additional changes in which the HSP objects are given the correct `output_index` argument, after rebasing.

**Comment by linucks** [comment]
Thanks for the helpful commands @bow.  It didn't go quite as smoothly as I'd have hoped (me and git don't get along, so I ended up with some conflicts during the merge), but think I've fixed that and have pushed some changes that should address everything.

**Comment by bow** [comment]
Ah, hmm...it looks like this was made into a merge commit instead of a rebase actually. It seems to have been caused by the local master being behind several commits of the upstream master.

Let's do it like this then: I'll close this merge request, rebase your changes on my local repo, then I'll open a new one that contains all the relevant commits rebased (keeping the history linear). These commits will still contain the same changes and the same author, only the hashes would be different, as you will see.

I'll also add an entry to `NEWS.rst` mentioning this changes. Just to confirm again, would you also like to have your name listed in `CONTRIB.rst`?

**Pull request closed without merging**